src/plugins/data/common/search/search_source/create_search_source.ts

```diff
@@ -33,18 +34,47 @@ export const createSearchSource = (
   indexPatterns: DataViewsContract,
   searchSourceDependencies: SearchSourceDependencies
 ) => {
-  const createFields = async (searchSourceFields: SerializedSearchSourceFields = {}) => {
+  let dataViewLazy: DataViewLazy | undefined;
+  const createFields = async (
+    searchSourceFields: SerializedSearchSourceFields = {},
+    useDataViewLazy = false
+  ) => {
     const { index, parent, ...restOfFields } = searchSourceFields;
     const fields: SearchSourceFields = {
       ...restOfFields,
     };
 
     // hydrating index pattern
     if (searchSourceFields.index) {
-      if (typeof searchSourceFields.index === 'string') {
-        fields.index = await indexPatterns.get(searchSourceFields.index);
+      if (!useDataViewLazy) {
+        if (typeof searchSourceFields.index === 'string') {
+          fields.index = await indexPatterns.get(searchSourceFields.index);
+        } else {
+          fields.index = await indexPatterns.create(searchSourceFields.index);
+        }
       } else {
-        fields.index = await indexPatterns.create(searchSourceFields.index);
+        if (typeof searchSourceFields.index === 'string') {
+          dataViewLazy = await indexPatterns.getDataViewLazy(searchSourceFields.index);
+          const dataView = new DataView({
+            spec: await dataViewLazy.toSpec(),
+            // field format functionality is not used within search source
+            fieldFormats: {} as FieldFormatsStartCommon,
+            shortDotsEnable: await searchSourceDependencies.dataViews.getShortDotsEnable(),
+            metaFields: await searchSourceDependencies.dataViews.getMetaFields(),
+          });
+
+          fields.index = dataView;
+        } else {
+          dataViewLazy = await indexPatterns.createDataViewLazy(searchSourceFields.index);
+          const dataView = new DataView({
+            spec: await dataViewLazy.toSpec(),
+            // field format functionality is not used within search source
+            fieldFormats: {} as FieldFormatsStartCommon,
+            shortDotsEnable: await searchSourceDependencies.dataViews.getShortDotsEnable(),
+            metaFields: await searchSourceDependencies.dataViews.getMetaFields(),
+          });
+          fields.index = dataView;
+        }
```

Comment:
For some reason using `isIndexName` in the conditional causes typescript to lose the more specific `searchSourceFields.index` type. Otherwise I've adopted these improvements.